presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/StreamPropertyDerivations.java

```diff
@@ -0,0 +1,566 @@
+                return StreamProperties.singleStream();
+            }
+
+            switch (node.getType()) {
+                case GATHER:
+                    return StreamProperties.singleStream();
+                case REPARTITION:
+                    if (node.getPartitionFunction().getPartitioningHandle().equals(FIXED_RANDOM_DISTRIBUTION)) {
+                        return new StreamProperties(FIXED, Optional.empty(), false);
+                    }
+                    return new StreamProperties(
+                            FIXED,
+                            Optional.of(
+                            node.getPartitionFunction().getPartitionFunctionArguments().stream()
+                                    .map(PartitionFunctionArgumentBinding::getColumn)
+                                    .collect(toImmutableList())),
+                            false);
+                case REPLICATE:
+                    return new StreamProperties(MULTIPLE, Optional.empty(), false);
+            }
+
+            throw new UnsupportedOperationException("not yet implemented");
+        }
+
+        //
+        // Nodes that rewrite and/or drop symbols
+        //
+
+        @Override
+        public StreamProperties visitProject(ProjectNode node, List<StreamProperties> inputProperties)
+        {
+            StreamProperties properties = Iterables.getOnlyElement(inputProperties);
+
+            // Only identity projected symbols are passed through
+            Map<Symbol, Symbol> identities = computeIdentityTranslations(node.getAssignments());
+            return properties.translate(column -> Optional.ofNullable(identities.get(column)));
+        }
+
+        private static Map<Symbol, Symbol> computeIdentityTranslations(Map<Symbol, Expression> assignments)
+        {
+            Map<Symbol, Symbol> inputToOutput = new HashMap<>();
+            for (Map.Entry<Symbol, Expression> assignment : assignments.entrySet()) {
+                if (assignment.getValue() instanceof QualifiedNameReference) {
+                    inputToOutput.put(Symbol.fromQualifiedName(((QualifiedNameReference) assignment.getValue()).getName()), assignment.getKey());
+                }
+            }
+            return inputToOutput;
+        }
+
+        @Override
+        public StreamProperties visitAggregation(AggregationNode node, List<StreamProperties> inputProperties)
+        {
+            StreamProperties properties = Iterables.getOnlyElement(inputProperties);
+
+            // Only grouped symbols projected symbols are passed through
+            return properties.translate(symbol -> node.getGroupBy().contains(symbol) ? Optional.of(symbol) : Optional.<Symbol>empty());
+        }
+
+        @Override
+        public StreamProperties visitTableFinish(TableFinishNode node, List<StreamProperties> inputProperties)
+        {
+            StreamProperties properties = Iterables.getOnlyElement(inputProperties);
+            // table commit does not output any symbols
```

Comment:
I don't think this statement is true. At a minimum, it outputs the count of rows added. It may not matter for properties, so I would just say that, instead.